File: examples/developer/guardrails.py
"""
Output Guardrails
==================
Demonstrate how to attach guardrails that filter or replace agent
output in real time. Two guardrail types are shown:

1. **Blocked terms** -- a simple deny-list that scrubs competitor names.
2. **Custom check** -- a callable that inspects each output chunk and
   substitutes a safe replacement when the check fails.

Requirements:
    pip install patter python-dotenv

Environment variables (.env):
    OPENAI_API_KEY      - OpenAI API key with Realtime access
    TWILIO_ACCOUNT_SID  - Twilio account SID
    TWILIO_AUTH_TOKEN   - Twilio auth token
    TWILIO_PHONE_NUMBER - Your Twilio phone number (E.164 format)
    WEBHOOK_URL         - Public URL where Twilio can reach this server
"""

import asyncio
import os

from dotenv import load_dotenv

load_dotenv()

from patter import Patter

# ── Initialise the Patter client ──────────────────────────────────────
phone = Patter(
    mode="local",
    openai_key=os.getenv("OPENAI_API_KEY"),
    twilio_sid=os.getenv("TWILIO_ACCOUNT_SID"),
    twilio_token=os.getenv("TWILIO_AUTH_TOKEN"),
    phone_number=os.getenv("TWILIO_PHONE_NUMBER"),
    webhook_url=os.getenv("WEBHOOK_URL"),
)

# ── Guardrail 1: block competitor mentions ────────────────────────────
phone.guardrail(
    name="no-competitors",
    blocked_terms=["CompetitorA", "CompetitorB"],
)

# ── Guardrail 2: prevent specific pricing disclosure ─────────────────
phone.guardrail(
    name="no-pricing",
    check=lambda text: "$" not in text,
    replacement="I can't discuss specific pricing on this call.",
)

# ── Define the voice agent with both guardrails applied ───────────────
agent = phone.agent(
    system_prompt=(
        "You are a sales representative for Acme Widgets. Answer "
        "questions about our products, features, and general pricing "
        "tiers (e.g. 'Starter', 'Pro', 'Enterprise'). Never mention "
        "competitor products by name and never quote exact dollar amounts."
    ),
    voice="nova",
    first_message="Hi there! I'm happy to tell you about Acme Widgets. What would you like to know?",
    guardrails=["no-competitors", "no-pricing"],
)

# ── Start the server ──────────────────────────────────────────────────
if __name__ == "__main__":
    asyncio.run(phone.serve(agent, port=8000))

File: examples/developer/guardrails.ts
/**
 * @file guardrails.ts
 * @description Voice agent with output guardrails.
 *
 * Demonstrates two guardrail types:
 *   1. Blocked-terms guardrail — prevents the agent from mentioning competitors.
 *   2. Custom-check guardrail — strips pricing information from responses.
 *
 * Guardrails run on every agent response before it reaches the caller, ensuring
 * compliance without modifying the system prompt.
 *
 * Prerequisites:
 *   - OPENAI_API_KEY, TWILIO_ACCOUNT_SID, TWILIO_AUTH_TOKEN in .env
 *
 * Usage:
 *   npx ts-node guardrails.ts
 */

import { Patter } from "getpatter";
import dotenv from "dotenv";

dotenv.config();

const phone = new Patter({
  mode: "local",
  openaiKey: process.env.OPENAI_API_KEY!,
  twilioSid: process.env.TWILIO_ACCOUNT_SID!,
  twilioToken: process.env.TWILIO_AUTH_TOKEN!,
  phoneNumber: process.env.TWILIO_PHONE_NUMBER!,
  webhookUrl: process.env.WEBHOOK_URL!,
});

const noCompetitors = Patter.guardrail({
  name: "no-competitors",
  blockedTerms: ["CompetitorA", "CompetitorB", "RivalCorp"],
});

const noPricing = Patter.guardrail({
  name: "no-pricing",
  check: (text: string) => !text.includes("$"),
  replacement:
    "I'm not able to share specific pricing over the phone. I can email you a detailed quote — would that work?",
});

const agent = phone.agent({
  systemPrompt: `You are a sales assistant for Acme Corp. Answer product questions,
highlight features, and help callers understand how Acme solutions fit their needs.
Never mention competitor products or share specific dollar amounts on the call.`,
  voice: "nova",
  firstMessage:
    "Welcome to Acme Corp! I'd love to help you find the right solution. What are you looking for?",
  guardrails: [noCompetitors, noPricing],
});

async function main(): Promise<void> {
  await phone.serve({ agent, port: 8000 });
  console.log("Guardrails agent listening on port 8000");
}

main();
